### PR TITLE
feat: Add ASGI root path parameter to Phoenix server

### LINKED ISSUE #2918: Add Env Parameter for Prefix for the app
**Is your feature request related to a problem? 
When trying to run phoenix behind load balancer with routing using endpoints it is impossible to do so.

For example web app: www.app.com having auth layer and then underlying services under routes /backend /phoenix and /* for frontend. It is not possible to change Phoenix routing path.

**Describe the solution you'd like**
Using env variable for example ROUTE_PREFIX or BASE_URL we could set to "/phoenix" and the all requests would go towards it, for example www.app.com/phoenix/graphql would point to phoenix backend or www.app.com/phoenix would point towards phoenix frontend.

**Describe alternatives you've considered**
Running phoenix service under separate domain name.

PS

Nothing to do with app.com, its just an example domain.

### PR DESCRIPTION
**Pull request by sbachstein**
Fixes #2918 

Allow for a root path parameter so that Phoenix can be hosted behind a proxy with a path prefix.

`Session` code has not been adjusted because the adjustment of the root path is probably not needed in this scenario.
`Client` and `Exporter` code have not been adjusted because they provide explicit `endpoint` parameters where a full URL can be given.

### CODE REVIEW

File: src/phoenix/config.py [hunk 1]
```diff
@@ -10,6 +10,7 @@
 ENV_PHOENIX_PORT = "PHOENIX_PORT"
 ENV_PHOENIX_GRPC_PORT = "PHOENIX_GRPC_PORT"
 ENV_PHOENIX_HOST = "PHOENIX_HOST"
+ENV_PHOENIX_ROOT_PATH = "PHOENIX_ROOT_PATH"
```
Comment (anticorrelator) [inline src/phoenix/config.py:13]: I think the changing env var name `PHOENIX_HOST_ROOT_PATH` is a bit better? We manage all of our configuration via environment variables and `PHOENIX_ROOT_PATH` might sound a little too generic

**Comment by anticorrelator** [comment]
Thanks so much for the contribution @sbachstein! This looks great, small request for a more specific config name

**Comment by RogerHYang** [comment]
The client needs to be updated, otherwise it'll have the wrong url.

e.g.
https://github.com/Arize-ai/phoenix/blob/5e8bc3dad2036b2febb1d343103d331723413a72/src/phoenix/session/client.py#L244

File: src/phoenix/config.py (hunk 1 above)
Comment (sbachstein) [reply src/phoenix/config.py:13]: Agreed, should we match the CLI parameter then to also be `--host_root_path` or should we stick with the `--root-path` parameter as it is known from uvicorn, for example?

**Comment by sbachstein** [comment]
> The client needs to be updated, otherwise it'll have the wrong url.
> 
> e.g.
> 
> https://github.com/Arize-ai/phoenix/blob/5e8bc3dad2036b2febb1d343103d331723413a72/src/phoenix/session/client.py#L244

The client provides the option to take the base path from the endpoint parameter (or environment variable). Do you think this is not sufficient and we need to, additionally, offer the option to specify HOST, PORT and ROOT_PATH?

File: src/phoenix/config.py (hunk 1 above)
Comment (sbachstein) [reply src/phoenix/config.py]: Went with the matching CLI parameter for now.

**Comment by RogerHYang** [comment]
The issue is that `urljoin` is currently absolute. See the four scenarios below

Screenshot 2024-05-15 at 7 40 16 AM

**Comment by sbachstein** [comment]
Very good catch, thank you for the hint. I adjusted all usages of `urljoin` to avoid the absolute paths.

**Comment by RogerHYang** [comment]
Thanks for the update. Can we make sure that `base_url` always ends with `/`, otherwise the `urljoin` is still not correct. 

`abc` below is gone because it does end with `/`

```pycon
>>> from urllib.parse import urljoin
>>> urljoin("http://xyz.com/abc", "rst")
'http://xyz.com/rst'
```

While we're at it, we could also add the root path as shown below.
```python
self._base_url = ( 
    endpoint
    or get_env_collector_endpoint()
    or urljoin(f"http://{host}:{get_env_port()}", get_env_host_root_path())
).rstrip("/") + "/"
```

https://github.com/Arize-ai/phoenix/blob/5e8bc3dad2036b2febb1d343103d331723413a72/src/phoenix/session/client.py#L58-L60

File: src/phoenix/server/main.py [hunk 2]
```diff
@@ -119,6 +120,7 @@ def _get_pid_file() -> Path:
     parser.add_argument("--export_path")
     parser.add_argument("--host", type=str, required=False)
     parser.add_argument("--port", type=int, required=False)
+    parser.add_argument("--host-root-path", type=str, required=False)
```
Comment (axiomofjoy) [inline src/phoenix/server/main.py:123]: Not sure we need the command line argument here in addition to the environment variable. We've been deprecating command line arguments such as `--enable-prometheus` that have environment variable equivalents if setting them via environment variables feels more natural, e.g., in Kubernetes. I think host root path probably falls into that category.

Comment (sbachstein) [reply src/phoenix/server/main.py]: Agreed, I removed the CLI parameter.

**Comment by sbachstein** [comment]
@RogerHYang Thank you for the hint again, that was kind of unexpected behavior for me, good to know!

File: tests/session/test_client.py [hunk 3]
```diff
@@ -18,6 +18,29 @@
 from phoenix.trace.trace_dataset import TraceDataset
 
 
+def test_base_path(monkeypatch: pytest.MonkeyPatch):
```
Comment (anticorrelator) [inline tests/session/test_client.py:21]: thank you for the test!

**Review by anticorrelator — APPROVED** [review]

**Review by RogerHYang — CHANGES_REQUESTED** [review]
I just tried this and I don't think the uvicorn implementation is correct.

At the minimum the traces endpoint is not advertised accurately.

Screenshot 2024-05-16 at 12 20 50 PM
Screenshot 2024-05-16 at 12 21 11 PM

Screenshot 2024-05-16 at 12 17 40 PM

Screenshot 2024-05-16 at 12 17 49 PM

Screenshot 2024-05-16 at 12 20 01 PM

**Comment by anticorrelator** [comment]
@sbachstein looks like the starlette app needs to be made aware of the root path too, thanks for checking this @RogerHYang

**Comment by sbachstein** [comment]
To me that is expected behavior. If the application is mounted at a root path with prefix by a proxy then the proxy would also make sure to rewrite the HTTP path and remove the root path from the request. That is also how I understand and have experienced the mechanisms of uvicorn/starlette root paths.

**Review by RogerHYang — APPROVED** [review]
Thanks, this is LGTM. Sorry for the hold up.

**Comment by sbachstein** [comment]
Thank you for the thorough review!

**Comment by axiomofjoy** [comment]
@sbachstein Thanks for making Phoenix better!

**Pull request merged**